feat: allow overriding `{host,bootstrap}-containers` entrypoint command

kasimeka opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

implements #593 

**Description of changes:**

- adds a `--command` option to `host-ctr` that overrides a container's `ProcessArgs` during creation
- adds a `command` field to the `bootstrap-containers` and `host-containers` apis that populates systemd service envs
- updates systemd services to pass the `--command` option to `host-ctr` for both `host-containers` and `bootstrap-containers`

**Testing done:**

- created an AMI from a build of this `core-kit` and validated that
  - `host-ctr run --source  328549459982.dkr.ecr.us-east-1.amazonaws.com/bottlerocket-bootstrap:v0.2.4 --container-id hysm --command 'sh,-c,echo "hello world" && sleep infinity'` works as expected and that containers launched with this command are accessible with `apiclient exec`
  - the control and admin containers launch normally with no changes to their default config

**work in progress:**

in the current state, im getting a "Json deserialize error: unknown field `command`" whenever i try to create a host-container with a command through the apiclient as follows
```shell
apiclient set -j '{ "settings": { "host-containers": { "mycontainer": {
  "enabled": true,
  "source":"328549459982.dkr.ecr.us-east-1.amazonaws.com/bottlerocket-bootstrap:v0.2.4",
  "command": ["sleep","infinity"],
  "user-data": "ZWNobyBoZW5sb28K" } } } }'
```
despite making this pr <https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/94> and updating all references to `bottlerocket-os/bottlerocket-settings-sdk` in `sources/Cargo.toml` to point at it instead. im not sure what's missing to make the apiserver pick up on the new model schema.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

kasimeka commented on packages/os/host-containers-toml:
```diff
@@ -17,5 +17,12 @@ superpowered = {{this.superpowered}}
+  "{{{this}}}",
+{{/each}}
+]
+{{/if_not_null}}
```
im not sure what this changes affects or how to test it, but it seems nothing's broken by it.

bcressey commented:
> I'm not sure what's missing to make the apiserver pick up on the new model schema.

In addition to the change here, you would also need to update the settings SDK dependencies in your fork / clone of `bottlerocket-os/bottlerocket`, or else the settings plugin that you compile won't know about the new field.

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -299,7 +301,13 @@ where
+        source,
+        &mode,
+        essential,
+        container_details.command.join(","),
```
nit: just for clarity, I'd probably do this with a separate binding:
```
let command = container_details.command.join(",");
```

kasimeka replied:
it conflicts with the `command` function, so getting rid of the binding was the simplest workaround. i can change every reference to the `command` function to `crate::command` if you prefer

bcressey replied:
Maybe `let entrypoint = ...` then?

kasimeka replied:
ive kept `let command =` and changed references to the `command` function to `crate::command`, so the field `command`'s name stays consistent everywhere. wdyt of this?

bcressey commented on sources/api/host-containers/src/main.rs:
```diff
@@ -360,7 +369,13 @@ where
+        source,
+        enabled,
+        superpowered,
+        image_details.command.join(","),
```
same nit as for bootstrap containers:
```
let command = image_details.command.join(",");
```

bcressey commented on sources/api/bootstrap-containers/src/main.rs:
```diff
@@ -299,7 +301,13 @@ where
+        source,
+        &mode,
+        essential,
+        container_details.command.join(","),
```
For overall safety against weird edge cases, like`,` as part of the command - might be good to borrow the `toml_encode` trick and write the vec out as a JSON string. Then `host-ctr` can deserialize from JSON back into an array.

kasimeka committed:
host-ctr: add --command flag for custom container entrypoint

Add --command flag that accepts a JSON array to override container entrypoint
using OCI's WithProcessArgs spec option.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

kasimeka committed:
host-containers: add support for custom entrypoint commands

Add command field to host container configuration and corresponding
template support to customize container entrypoints. Implements CTR_COMMAND
environment variable in service files that is passed to host-ctr.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

kasimeka committed:
bootstrap-containers: add support for custom entrypoint commands

Add command field to bootstrap container configuration and corresponding
template support to customize container entrypoints. Implements CTR_COMMAND
environment variable in service files that is passed to host-ctr.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

kasimeka commented:
@bcressey  `host-ctr` works fine with a json passed to `--command` but im still unable to test the change through `apiclient`. here's my current `bottlerocket-os/bottlerocket` branch <https://github.com/kasimeka/bottlerocket/tree/13d98df50d243cfb1e2d345edf2afab1bf9c60f3> (my settings-sdk PR was merged and tagged), any idea what else is missing?
also please let me know if it's fine to rebase on `develop` and squash the commits added in response to review comments.

ytsssun commented:
> `host-ctr` works fine with a json passed to `--command` but im still unable to test the change through `apiclient`. here's my current `bottlerocket-os/bottlerocket` branch https://github.com/kasimeka/bottlerocket/tree/13d98df50d243cfb1e2d345edf2afab1bf9c60f3 (my settings-sdk PR was merged and tagged), any idea what else is missing? 

Hi @kasimeka , I can help with that. I am building a custom core-kit to test it out. I will let you know how that goes. We are targeting to merge this change for the upcoming core-kit release. Do you mind if I directly push changes to your branch?

> also please let me know if it's fine to rebase on `develop` and squash the commits added in response to review comments.

Yes we need to squash the commits to clean up the history and organize them into modular commits.

bcressey commented:
Changes LGTM, thanks! Should be good to go once the rebase is done.

ytsssun commented:
I was able to verify the change works.

I ran the custom host container via

```
apiclient set -j '{ "settings": { "host-containers": { "mycontainer": {
  "enabled": true,
  "source":"328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-bootstrap:v0.2.5",
  "command": ["sleep","infinity"],
  "user-data": "ZWNobyBoZW5sb28K" } } } }'
```
And it is showing us in the api setting
```
bash-5.1# apiclient get settings.host-containers
{
  "settings": {
    "host-containers": {
      "admin": {
        ...
      },
      "control": {
        ...
      },
      "mycontainer": {
        "command": [
          "sleep",
          "infinity"
        ],
        "enabled": true,
        "source": "328549459982.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-bootstrap:v0.2.5",
        "user-data": "ZWNobyBoZW5sb28K"
      }
    }
  }
}
```

Verified the container is running via ctr
```
bash-5.1# ctr -a /run/host-containerd/containerd.sock c ls
CONTAINER      IMAGE                                                                                  RUNTIME
admin          ecr.aws/arn:aws:ecr:us-west-2:328549459982:repository/bottlerocket-admin:v0.12.2       io.containerd.runc.v2
control        ecr.aws/arn:aws:ecr:us-west-2:328549459982:repository/bottlerocket-control:v0.8.7      io.containerd.runc.v2
mycontainer    ecr.aws/arn:aws:ecr:us-west-2:328549459982:repository/bottlerocket-bootstrap:v0.2.5    io.containerd.runc.v2
```

This also confirms that host-ctr change worked as that indirectly triggers host-ctr to launch the container.

I further tried using `apiclient exec mycontainer sh`, and verified that I am inside the custom container:
```
bash-5.1# apiclient exec mycontainer sh
sh-5.2# cat /etc/os-release
NAME="Amazon Linux"
VERSION="2023"
ID="amzn"
ID_LIKE="fedora"
VERSION_ID="2023"
PLATFORM_ID="platform:al2023"
PRETTY_NAME="Amazon Linux 2023.8.20250804"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2023"
HOME_URL="https://aws.amazon.com/linux/amazon-linux-2023/"
DOCUMENTATION_URL="https://docs.aws.amazon.com/linux/"
SUPPORT_URL="https://aws.amazon.com/premiumsupport/"
BUG_REPORT_URL="https://github.com/amazonlinux/amazon-linux-2023"
VENDOR_NAME="AWS"
VENDOR_URL="https://aws.amazon.com/"
SUPPORT_END="2029-06-30"
```

> despite making this pr https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/94 and updating all references to bottlerocket-os/bottlerocket-settings-sdk in sources/Cargo.toml to point at it instead. im not sure what's missing to make the apiserver pick up on the new model schema.


I simply bumped the core-kit and settings-sdk versions in my AMI build. My guess why your testing did not work is, potentially either your AMI build or your core-kit build was using come cached changes and did not take your new settings-sdk update. 

For core-kit local testing, I always bump the patch version of the kit and I cleaned up the build cache in BoB to build a fresh AMI.

ytsssun commented:
Force pushed the changes to [kasimeka:add-ctr-command](https://github.com/kasimeka/bottlerocket-core-kit/tree/add-ctr-command) branch with cleaned-up commits.

ytsssun commented:
I also attempted a migration test as this change would require a new migration behavior where we ensure the downgrade of the AMI will support removing all the container settings for `command` fields - that also includes the custom containers.

This PR has the migration and the testing I did - https://github.com/bottlerocket-os/bottlerocket/pull/4636
